crates/chain-state/src/notifications.rs

```diff
@@ -211,6 +212,74 @@ impl<T: Clone + Sync + Send + 'static> Stream for ForkChoiceStream<T> {
     }
 }
 
+/// A stream that emits canonical chain notifications only when blocks are finalized.
+#[derive(Debug)]
+#[pin_project]
+pub(crate) struct FinalizedNotificationStream<N: NodePrimitives> {
+    #[pin]
+    canon_stream: CanonStateNotificationStream<N>,
+    #[pin]
+    finalized_stream: ForkChoiceStream<SealedHeader<alloy_consensus::Header>>,
+    // Buffer for the most recent canonical notification
+    buffered_notification: Option<CanonStateNotification<N>>,
```

Comment:
it's also possible that the finalized messages arrives early so we need to account for that as well